### PR TITLE
Add check for parents in Quotes (#19842)

### LINKED ISSUE #19842: Regression for frugalmechanic/fm-serializer in quotes-reflect + -Xcheck-macros
Based on OpenCB failure of `frugalmechanic/fm-serializer` - [build logs](https://github.com/VirtusLab/community-build3/actions/runs/8026132604/job/21929132671) 
Under -Xcheck-macros flag we get an error about different sets of parents - the parents of tree don't include `class Object` which is present in parents of symbol being created. Removing the `class Object` from list of parents leads to assertion error in the compiler, becouse it requires the first param to be a class, not a trait.

## Compiler version
3.4.nightly 
Last good release: 3.4.1-RC1-bin-20240210-6efcdba-NIGHTLY
First bad release: 3.4.1-RC1-bin-20240212-c529a48-NIGHTLY
Bisect points to aa5492ffd4559301f26b1bb1e6da9642700c7999

## Minimized code

```scala
// main.scala
trait Serializer[@specialized T]// extends RawSerializer[T] with NestedSerializer[T] with FieldSerializer[T]
object Serializer{
    implicit inline def implicitMakeSerializer[T]: Serializer[T] = ${ Macros.makeSerializer[T] }
}
case class ValidationCls(string: String)

@main def Test = summon[Serializer[ValidationCls]]
```

```Scala
// macros.scala
//> using options -Xcheck-macros
import scala.annotation.{experimental, targetName}
import scala.quoted.*
import scala.util.Try

object Macros {
  def makeSerializer[T: Type](using Quotes): Expr[Serializer[T]] = {
    import quotes.reflect.*

    val tpe: TypeRepr = TypeRepr.of[T]
    val name: String = Symbol.freshName("objectSerializer")

    val modSym: Symbol = Symbol.newModule(
      Symbol.spliceOwner,
      name,
      Flags.Implicit,
      Flags.EmptyFlags,
      // Without TypeRep.of[Object] it would fail with java.lang.AssertionError: assertion failed: First parent must be a class
      List(TypeRepr.of[Object], TypeRepr.of[Serializer[T]]),
      _ => Nil,
      Symbol.noSymbol
    )

    val (modValDef: ValDef, modClassDef: ClassDef) =
      ClassDef.module(modSym, List(TypeTree.of[Serializer[T]]), Nil)

    Block(List(modValDef, modClassDef), Ref(modSym)).asExprOf[Serializer[T]]
  }
}
```

## Output

```scala
-- Error: /Users/wmazur/projects/dotty/bisect/main.scala:7:50 ------------------
7 |@main def Test = summon[Serializer[ValidationCls]]
  |                                                  ^
  |Malformed tree was found while expanding macro with -Xcheck-macros.
  |The tree does not conform to the compiler's tree invariants.
  |
  |Macro was:
  |scala.quoted.runtime.Expr.splice[Serializer[ValidationCls]](((contextual$2: scala.quoted.Quotes) ?=> Macros.makeSerializer[ValidationCls](scala.quoted.Type.of[ValidationCls](contextual$2), contextual$2)))
  |
  |The macro returned:
  |{
  |  object objectSerializer$macro$1 extends Serializer[ValidationCls] { this: objectSerializer$macro$1.type =>
  |    
  |  }
  |  objectSerializer$macro$1
  |}
  |
  |Error:
  |assertion failed: Parents of class symbol differs from the parents in the tree for object objectSerializer$macro$1
  |
  |Parents in symbol: [class Object, trait Serializer]
  |Parents in tree: [trait Serializer]
  |
  |
  |stacktrace available when compiling with `-Ydebug`
  |-----------------------------------------------------------------------------
  |Inline stack trace
  |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  |This location contains code that was inlined from main.scala:3
3 |    implicit inline def implicitMakeSerializer[T]: Serializer[T] = ${ Macros.makeSerializer[T] }
  |                                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   -----------------------------------------------------------------------------
1 error found
Compilation failed
```

## Expectation
Probably should not couse comilation error.

### PR DESCRIPTION
**Pull request by hamzaremmal**
Fix #19842

### CODE REVIEW

**Review by nicolasstucki — APPROVED** [review]

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala [hunk 1]
```diff
@@ -253,6 +253,7 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
         (cdef.name.toString, cdef.constructor, cdef.parents, cdef.self, rhs.body)
 
       def module(module: Symbol, parents: List[Tree /* Term | TypeTree */], body: List[Statement]): (ValDef, ClassDef) = {
+        xCheckMacroAssert(parents.nonEmpty && !parents.head.tpe.dealias.classSymbol.is(dotc.core.Flags.Trait), "First parent must be a class")
```
Comment (nicolasstucki) [inline compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:256]: Here we have the symbol and the parents. We could use the TreeChecker logic to check that all parents align.

**Review by nicolasstucki — CHANGES_REQUESTED** [review]

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala (hunk 1 above)
Comment (hamzaremmal) [reply compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:256]: We can, but it's going to be a much stronger test. Should we do that ? Can we do that without other restrictions ?

Comment (nicolasstucki) [reply compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:256]: > Should we do that ?

It is better to have a stronger test here. This way, if the user makes a mistake they will know exactly where it happened and how to fix it. When the failure is in -Ycheck they still have to trace back to find where this issue happened.

> Can we do that without other restrictions ?

What kind of restriction?

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala [hunk 2]
```diff
@@ -253,6 +253,12 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
...
+        xCheckMacroAssert(
```
Comment (nicolasstucki) [inline compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:257]: We should factor out the common logic with https://github.com/scala/scala3/pull/18935/files#diff-0819ff2757ddb9b8950698b5e502c765cc74ff4ddbd0a5a729b71d9e3eb8b217R554-R563

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala [hunk 3]
```diff
@@ -253,6 +253,12 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
         (cdef.name.toString, cdef.constructor, cdef.parents, cdef.self, rhs.body)
 
       def module(module: Symbol, parents: List[Tree /* Term | TypeTree */], body: List[Statement]): (ValDef, ClassDef) = {
+        xCheckMacroAssert(parents.nonEmpty && !parents.head.tpe.dealias.classSymbol.is(dotc.core.Flags.Trait), "First parent must be a class")
```
Comment (nicolasstucki) [inline compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:256]: This test seems unnecessary, if the parents align with the symbol parents we should never have a trait as the first parent.

**Review by nicolasstucki — CHANGES_REQUESTED** [review]

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala [hunk 4]
```diff
@@ -253,6 +254,7 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
         (cdef.name.toString, cdef.constructor, cdef.parents, cdef.self, rhs.body)
 
       def module(module: Symbol, parents: List[Tree /* Term | TypeTree */], body: List[Statement]): (ValDef, ClassDef) = {
+        TreeChecker.checkParents(module.moduleClass.asClass, parents, xCheckMacroAssert)
```
Comment (nicolasstucki) [inline compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:257]:
```suggestion
        if xCheckMacro then TreeChecker.checkParents(module.moduleClass.asClass, parents)
```

File: compiler/src/dotty/tools/dotc/transform/TreeChecker.scala [hunk 5]
```diff
@@ -198,6 +198,17 @@ object TreeChecker {
     }
   }.apply(tp0)
 
+  def checkParents(sym: ClassSymbol, parents: List[tpd.Tree], check: (Boolean, String) => Unit)(using Context): Unit =
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/transform/TreeChecker.scala:201]:
```suggestion
  def checkParents(sym: ClassSymbol, parents: List[tpd.Tree])(using Context): Unit =
```

File: compiler/src/dotty/tools/dotc/transform/TreeChecker.scala [hunk 6]
```diff
@@ -198,6 +198,17 @@ object TreeChecker {
...
+    val symbolParents = sym.classInfo.parents.map(_.dealias.typeSymbol)
+    val treeParents = parents.map(_.tpe.dealias.typeSymbol)
+    check(symbolParents == treeParents,
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/transform/TreeChecker.scala:204]:
```suggestion
    assert(symbolParents == treeParents,
```

**Review by nicolasstucki — CHANGES_REQUESTED** [review]

**Review by nicolasstucki — APPROVED** [review]

**Pull request merged**